File: gestores/gestor_personas.py
from __future__ import annotations

from datetime import date
from decimal import Decimal
from typing import Any, Iterable

from modelo_datos import Administrativo, Estudiante, EstadoAcademico, Persona, Profesor


class ErrorPersona(ValueError):
    """Error de integridad o validación de personas."""


class GestorPersonas:
    def __init__(
        self,
        personas: list[Persona],
        estudiantes: list[Estudiante] | None = None,
        profesores: list[Profesor] | None = None,
        administrativos: list[Administrativo] | None = None,
        programas: Iterable[Any] | None = None,
        planes_estudio: Iterable[Any] | None = None,
    ) -> None:
        self.personas = personas
        self.estudiantes = estudiantes if estudiantes is not None else []
        self.profesores = profesores if profesores is not None else []
        self.administrativos = administrativos if administrativos is not None else []
        self.programas = list(programas or [])
        self.planes_estudio = list(planes_estudio or [])

    def crear_persona(self, persona: Persona) -> Persona:
        if persona.idPersona is None:
            persona.idPersona = self._siguiente_id(self.personas, "idPersona")
        if any(item.idPersona == persona.idPersona for item in self.personas):
            raise ErrorPersona(f"Ya existe una persona con ID {persona.idPersona}")
        if not persona.numeroDocumento:
            raise ErrorPersona("El número de documento es obligatorio")
        if any(item.numeroDocumento == persona.numeroDocumento for item in self.personas):
            raise ErrorPersona(f"Ya existe el documento {persona.numeroDocumento}")
        persona.fechaRegistro = persona.fechaRegistro or date.today()
        persona.estado = persona.estado or "ACTIVO"
        self.personas.append(persona)
        return persona

    def crear_estudiante(self, estudiante: Estudiante) -> Estudiante:
        self._persona_existente(estudiante.idPersona)
        if estudiante.idEstudiante is None:
            estudiante.idEstudiante = self._siguiente_id(self.estudiantes, "idEstudiante")
        self._id_unico(self.estudiantes, "idEstudiante", estudiante.idEstudiante, "estudiante")
        if any(item.codigoEstudiante == estudiante.codigoEstudiante for item in self.estudiantes if estudiante.codigoEstudiante):
            raise ErrorPersona(f"Ya existe el código de estudiante {estudiante.codigoEstudiante}")
        self._referencia_existente(self.programas, "idPrograma", estudiante.idPrograma, "programa")
        self._referencia_existente(self.planes_estudio, "idPlanEstudio", estudiante.idPlanEstudio, "plan de estudio")
        estudiante.estadoAcademico = estudiante.estadoAcademico or EstadoAcademico.ACTIVO
        estudiante.estado = estudiante.estado or "ACTIVO"
        self.estudiantes.append(estudiante)
        return estudiante

    def crear_profesor(self, profesor: Profesor) -> Profesor:
        self._persona_existente(profesor.idPersona)
        if profesor.idProfesor is None:
            profesor.idProfesor = self._siguiente_id(self.profesores, "idProfesor")
        self._id_unico(self.profesores, "idProfesor", profesor.idProfesor, "profesor")
        if any(item.codigoProfesor == profesor.codigoProfesor for item in self.profesores if profesor.codigoProfesor):
            raise ErrorPersona(f"Ya existe el código de profesor {profesor.codigoProfesor}")
        # Validación B5: Evitar registrar múltiples posgrados por profesor
        # El profesor ya tiene un nivel de posgrado reconocido, no se permite otro
        posgrado_existente = next(
            (p for p in self.profesores if p.idProfesor == profesor.idProfesor),
            None,
        )
        if posgrado_existente and posgrado_existente.nivelPosgradoReconocido is not None:
            raise ErrorPersona(
                f"El profesor con ID {profesor.idProfesor} ya tiene un posgrado reconocido "
                f"('{posgrado_existente.nivelPosgradoReconocido}'), no se pueden registrar múltiples posgrados"
            )
        profesor.estado = profesor.estado or "ACTIVO"
        self.profesores.append(profesor)
        return profesor

    def crear_administrativo(self, administrativo: Administrativo) -> Administrativo:
        self._persona_existente(administrativo.idPersona)
        if administrativo.idAdministrativo is None:
            administrativo.idAdministrativo = self._siguiente_id(self.administrativos, "idAdministrativo")
        self._id_unico(self.administrativos, "idAdministrativo", administrativo.idAdministrativo, "administrativo")
        if any(item.codigoEmpleado == administrativo.codigoEmpleado for item in self.administrativos if administrativo.codigoEmpleado):
            raise ErrorPersona(f"Ya existe el código de empleado {administrativo.codigoEmpleado}")
        administrativo.estado = administrativo.estado or "ACTIVO"
        self.administrativos.append(administrativo)
        return administrativo

    def buscar_persona_por_documento(self, numero_documento: str) -> Persona | None:
        return next((item for item in self.personas if item.numeroDocumento == numero_documento), None)

    def buscar_estudiante_por_codigo(self, codigo: str) -> Estudiante | None:
        return next((item for item in self.estudiantes if item.codigoEstudiante == codigo), None)

    def buscar_profesor_por_codigo(self, codigo: str) -> Profesor | None:
        return next((item for item in self.profesores if item.codigoProfesor == codigo), None)

    def listar_estudiantes_por_programa(self, id_programa: int) -> list[Estudiante]:
        return [item for item in self.estudiantes if item.idPrograma == id_programa]

    def listar_profesores_por_programa(self, id_programa: int) -> list[Profesor]:
        return [item for item in self.profesores if item.idProgramaPrincipal == id_programa]

    def modificar_persona(self, id_persona: int, **cambios: Any) -> Persona:
        persona = self._buscar(self.personas, "idPersona", id_persona, "persona")
        if "numeroDocumento" in cambios and cambios["numeroDocumento"] != persona.numeroDocumento:
            if any(item.numeroDocumento == cambios["numeroDocumento"] for item in self.personas):
                raise ErrorPersona(f"Ya existe el documento {cambios['numeroDocumento']}")
        self._actualizar(persona, cambios)
        return persona

    def desactivar_persona(self, id_persona: int) -> Persona:
        persona = self._buscar(self.personas, "idPersona", id_persona, "persona")
        persona.estado = "INACTIVO"
        return persona

    def desactivar_estudiante(self, id_estudiante: int) -> Estudiante:
        estudiante = self._buscar(self.estudiantes, "idEstudiante", id_estudiante, "estudiante")
        estudiante.estado = "INACTIVO"
        return estudiante

    def desactivar_profesor(self, id_profesor: int) -> Profesor:
        profesor = self._buscar(self.profesores, "idProfesor", id_profesor, "profesor")
        profesor.estado = "INACTIVO"
        return profesor

    def desactivar_administrativo(self, id_administrativo: int) -> Administrativo:
        administrativo = self._buscar(self.administrativos, "idAdministrativo", id_administrativo, "administrativo")
        administrativo.estado = "INACTIVO"
        return administrativo

    def _persona_existente(self, id_persona: int | None) -> Persona:
        if id_persona is None:
            raise ErrorPersona("El rol debe estar asociado a una persona")
        return self._buscar(self.personas, "idPersona", id_persona, "persona")

    @staticmethod
    def _referencia_existente(elementos: Iterable[Any], campo: str, valor: int | None, nombre: str) -> None:
        if valor is not None and not any(getattr(item, campo, None) == valor for item in elementos):
            raise ErrorPersona(f"No existe el {nombre} con ID {valor}")

    @staticmethod
    def _id_unico(elementos: Iterable[Any], campo: str, valor: int | None, nombre: str) -> None:
        if any(getattr(item, campo, None) == valor for item in elementos):
            raise ErrorPersona(f"Ya existe un {nombre} con ID {valor}")

    @staticmethod
    def _buscar(elementos: Iterable[Any], campo: str, valor: Any, nombre: str) -> Any:
        elemento = next((item for item in elementos if getattr(item, campo, None) == valor), None)
        if elemento is None:
            raise ErrorPersona(f"No existe el {nombre} con ID {valor}")
        return elemento

    @staticmethod
    def _actualizar(objeto: Any, cambios: dict[str, Any]) -> None:
        campos = set(getattr(objeto, "__dataclass_fields__", {}))
        desconocidos = set(cambios) - campos
        if desconocidos:
            raise ErrorPersona(f"Campos no válidos: {', '.join(sorted(desconocidos))}")
        for nombre, valor in cambios.items():
            setattr(objeto, nombre, valor)

    @staticmethod
    def _siguiente_id(elementos: Iterable[Any], campo: str) -> int:
        return max((getattr(item, campo) or 0 for item in elementos), default=0) + 1

File: cpp/gestores/gestor_personas.cpp
#include "gestor_personas.h"
#include <ctime>
#include <iomanip>
#include <sstream>

namespace pita {

static std::string fecha_hoy() {
    std::time_t t = std::time(nullptr);
    std::tm tm = *std::localtime(&t);
    std::ostringstream oss;
    oss << std::put_time(&tm, "%Y-%m-%d");
    return oss.str();
}

GestorPersonas::GestorPersonas(
    ListaEnlazada<Persona>& personas,
    ListaEnlazada<Estudiante>& estudiantes,
    ListaEnlazada<Profesor>& profesores,
    ListaEnlazada<Administrativo>& administrativos,
    ListaEnlazada<ProgramaAcademico>& programas,
    ListaEnlazada<PlanEstudio>& planesEstudio
) : personas(personas),
    estudiantes(estudiantes),
    profesores(profesores),
    administrativos(administrativos),
    programas(programas),
    planesEstudio(planesEstudio) {}

int GestorPersonas::siguienteIdPersona() {
    int maxId = 0;
    for (const auto& p : personas) {
        if (p.idPersona.has_value() && *p.idPersona > maxId) {
            maxId = *p.idPersona;
        }
    }
    return maxId + 1;
}

int GestorPersonas::siguienteIdEstudiante() {
    int maxId = 0;
    for (const auto& e : estudiantes) {
        if (e.idEstudiante.has_value() && *e.idEstudiante > maxId) {
            maxId = *e.idEstudiante;
        }
    }
    return maxId + 1;
}

int GestorPersonas::siguienteIdProfesor() {
    int maxId = 0;
    for (const auto& p : profesores) {
        if (p.idProfesor.has_value() && *p.idProfesor > maxId) {
            maxId = *p.idProfesor;
        }
    }
    return maxId + 1;
}

int GestorPersonas::siguienteIdAdministrativo() {
    int maxId = 0;
    for (const auto& a : administrativos) {
        if (a.idAdministrativo.has_value() && *a.idAdministrativo > maxId) {
            maxId = *a.idAdministrativo;
        }
    }
    return maxId + 1;
}

void GestorPersonas::verificarPersonaExistente(std::optional<int> idPersona) {
    if (!idPersona.has_value()) {
        throw ErrorPersona("El rol debe estar asociado a una persona");
    }
    for (const auto& p : personas) {
        if (p.idPersona.has_value() && *p.idPersona == *idPersona) {
            return;
        }
    }
    throw ErrorPersona("No existe la persona con ID " + std::to_string(*idPersona));
}

Persona& GestorPersonas::crearPersona(Persona persona) {
    if (!persona.idPersona.has_value()) {
        persona.idPersona = siguienteIdPersona();
    } else {
        for (const auto& p : personas) {
            if (p.idPersona.has_value() && *p.idPersona == *persona.idPersona) {
                throw ErrorPersona("Ya existe una persona con ID " + std::to_string(*persona.idPersona));
            }
        }
    }

    if (!persona.numeroDocumento.has_value() || persona.numeroDocumento->empty()) {
        throw ErrorPersona("El numero de documento es obligatorio");
    }

    for (const auto& p : personas) {
        if (p.numeroDocumento.has_value() && *p.numeroDocumento == *persona.numeroDocumento) {
            throw ErrorPersona("Ya existe el documento " + *persona.numeroDocumento);
        }
    }

    if (!persona.fechaRegistro.has_value() || persona.fechaRegistro->empty()) {
        persona.fechaRegistro = fecha_hoy();
    }
    if (!persona.estado.has_value() || persona.estado->empty()) {
        persona.estado = "ACTIVO";
    }

    personas.push_back(std::move(persona));
    return personas.back();
}

Estudiante& GestorPersonas::crearEstudiante(Estudiante estudiante) {
    verificarPersonaExistente(estudiante.idPersona);

    if (!estudiante.idEstudiante.has_value()) {
        estudiante.idEstudiante = siguienteIdEstudiante();
    } else {
        for (const auto& e : estudiantes) {
            if (e.idEstudiante.has_value() && *e.idEstudiante == *estudiante.idEstudiante) {
                throw ErrorPersona("Ya existe un estudiante con ID " + std::to_string(*estudiante.idEstudiante));
            }
        }
    }

    if (estudiante.codigoEstudiante.has_value() && !estudiante.codigoEstudiante->empty()) {
        for (const auto& e : estudiantes) {
            if (e.codigoEstudiante.has_value() && *e.codigoEstudiante == *estudiante.codigoEstudiante) {
                throw ErrorPersona("Ya existe el codigo de estudiante " + *estudiante.codigoEstudiante);
            }
        }
    }

    if (estudiante.idPrograma.has_value()) {
        bool existeProg = false;
        for (const auto& pr : programas) {
            if (pr.idPrograma.has_value() && *pr.idPrograma == *estudiante.idPrograma) {
                existeProg = true;
                break;
            }
        }
        if (!existeProg) {
            throw ErrorPersona("No existe el programa con ID " + std::to_string(*estudiante.idPrograma));
        }
    }

    if (estudiante.idPlanEstudio.has_value()) {
        bool existePlan = false;
        for (const auto& pl : planesEstudio) {
            if (pl.idPlanEstudio.has_value() && *pl.idPlanEstudio == *estudiante.idPlanEstudio) {
                existePlan = true;
                break;
            }
        }
        if (!existePlan) {
            throw ErrorPersona("No existe el plan de estudio con ID " + std::to_string(*estudiante.idPlanEstudio));
        }
    }

    if (!estudiante.estadoAcademico.has_value()) {
        estudiante.estadoAcademico = EstadoAcademico::ACTIVO;
    }
    if (!estudiante.estado.has_value() || estudiante.estado->empty()) {
        estudiante.estado = "ACTIVO";
    }

    estudiantes.push_back(std::move(estudiante));
    return estudiantes.back();
}

Profesor& GestorPersonas::crearProfesor(Profesor profesor) {
    verificarPersonaExistente(profesor.idPersona);

    if (!profesor.idProfesor.has_value()) {
        profesor.idProfesor = siguienteIdProfesor();
    } else {
        for (const auto& p : profesores) {
            if (p.idProfesor.has_value() && *p.idProfesor == *profesor.idProfesor) {
                throw ErrorPersona("Ya existe un profesor con ID " + std::to_string(*profesor.idProfesor));
            }
        }
    }

    if (profesor.codigoProfesor.has_value() && !profesor.codigoProfesor->empty()) {
        for (const auto& p : profesores) {
            if (p.codigoProfesor.has_value() && *p.codigoProfesor == *profesor.codigoProfesor) {
                throw ErrorPersona("Ya existe el codigo de profesor " + *profesor.codigoProfesor);
            }
        }
    }

    // Regla B5: Evitar registrar múltiples posgrados si ya tiene uno reconocido
    for (const auto& p : profesores) {
        if (p.idProfesor.has_value() && *p.idProfesor == *profesor.idProfesor) {
            if (p.nivelPosgradoReconocido.has_value() && !p.nivelPosgradoReconocido->empty()) {
                throw ErrorPersona("El profesor con ID " + std::to_string(*profesor.idProfesor) +
                                   " ya tiene un posgrado reconocido ('" + *p.nivelPosgradoReconocido +
                                   "'), no se pueden registrar multiples posgrados");
            }
        }
    }

    if (!profesor.estado.has_value() || profesor.estado->empty()) {
        profesor.estado = "ACTIVO";
    }

    profesores.push_back(std::move(profesor));
    return profesores.back();
}

Administrativo& GestorPersonas::crearAdministrativo(Administrativo administrativo) {
    verificarPersonaExistente(administrativo.idPersona);

    if (!administrativo.idAdministrativo.has_value()) {
        administrativo.idAdministrativo = siguienteIdAdministrativo();
    } else {
        for (const auto& a : administrativos) {
            if (a.idAdministrativo.has_value() && *a.idAdministrativo == *administrativo.idAdministrativo) {
                throw ErrorPersona("Ya existe un administrativo con ID " + std::to_string(*administrativo.idAdministrativo));
            }
        }
    }

    if (administrativo.codigoEmpleado.has_value() && !administrativo.codigoEmpleado->empty()) {
        for (const auto& a : administrativos) {
            if (a.codigoEmpleado.has_value() && *a.codigoEmpleado == *administrativo.codigoEmpleado) {
                throw ErrorPersona("Ya existe el codigo de empleado " + *administrativo.codigoEmpleado);
            }
        }
    }

    if (!administrativo.estado.has_value() || administrativo.estado->empty()) {
        administrativo.estado = "ACTIVO";
    }

    administrativos.push_back(std::move(administrativo));
    return administrativos.back();
}

Persona* GestorPersonas::buscarPersonaPorDocumento(const std::string& numeroDocumento) {
    for (auto& p : personas) {
        if (p.numeroDocumento.has_value() && *p.numeroDocumento == numeroDocumento) {
            return &p;
        }
    }
    return nullptr;
}

Estudiante* GestorPersonas::buscarEstudiantePorCodigo(const std::string& codigo) {
    for (auto& e : estudiantes) {
        if (e.codigoEstudiante.has_value() && *e.codigoEstudiante == codigo) {
            return &e;
        }
    }
    return nullptr;
}

Profesor* GestorPersonas::buscarProfesorPorCodigo(const std::string& codigo) {
    for (auto& p : profesores) {
        if (p.codigoProfesor.has_value() && *p.codigoProfesor == codigo) {
            return &p;
        }
    }
    return nullptr;
}

ListaEnlazada<Estudiante> GestorPersonas::listarEstudiantesPorPrograma(int idPrograma) const {
    ListaEnlazada<Estudiante> res;
    for (const auto& e : estudiantes) {
        if (e.idPrograma.has_value() && *e.idPrograma == idPrograma) {
            res.push_back(e);
        }
    }
    return res;
}

ListaEnlazada<Profesor> GestorPersonas::listarProfesoresPorPrograma(int idPrograma) const {
    ListaEnlazada<Profesor> res;
    for (const auto& p : profesores) {
        if (p.idProgramaPrincipal.has_value() && *p.idProgramaPrincipal == idPrograma) {
            res.push_back(p);
        }
    }
    return res;
}

Persona& GestorPersonas::desactivarPersona(int idPersona) {
    for (auto& p : personas) {
        if (p.idPersona.has_value() && *p.idPersona == idPersona) {
            p.estado = "INACTIVO";
            return p;
        }
    }
    throw ErrorPersona("No existe la persona con ID " + std::to_string(idPersona));
}

Estudiante& GestorPersonas::desactivarEstudiante(int idEstudiante) {
    for (auto& e : estudiantes) {
        if (e.idEstudiante.has_value() && *e.idEstudiante == idEstudiante) {
            e.estado = "INACTIVO";
            return e;
        }
    }
    throw ErrorPersona("No existe el estudiante con ID " + std::to_string(idEstudiante));
}

Profesor& GestorPersonas::desactivarProfesor(int idProfesor) {
    for (auto& p : profesores) {
        if (p.idProfesor.has_value() && *p.idProfesor == idProfesor) {
            p.estado = "INACTIVO";
            return p;
        }
    }
    throw ErrorPersona("No existe el profesor con ID " + std::to_string(idProfesor));
}

Administrativo& GestorPersonas::desactivarAdministrativo(int idAdministrativo) {
    for (auto& a : administrativos) {
        if (a.idAdministrativo.has_value() && *a.idAdministrativo == idAdministrativo) {
            a.estado = "INACTIVO";
            return a;
        }
    }
    throw ErrorPersona("No existe el administrativo con ID " + std::to_string(idAdministrativo));
}

} // namespace pita
